mmv1/templates/validator/mappers/mappers.go.erb

```diff
@@ -0,0 +1,211 @@
+// The incoming asset will either be an asset that was created/updated or deleted.
+type mergeFunc func(existing, incoming Asset) Asset
+
+// mapper pairs related conversion/merging functions.
+type mapper struct {
+	convert           convertFunc // required
+	fetch             fetchFunc   // optional
+	mergeCreateUpdate mergeFunc   // optional
+	mergeDelete       mergeFunc   // optional
+}
+
+// mappers maps terraform resource types (i.e. `google_project`) into
+// a slice of mapperFuncs.
+//
+// Modelling of relationships:
+// terraform resources to CAI assets as []mapperFuncs:
+// 1:1 = [mapper{convert: convertAbc}]                  (len=1)
+// 1:N = [mapper{convert: convertAbc}, ...]             (len=N)
+// N:1 = [mapper{convert: convertAbc, merge: mergeAbc}] (len=1)
+func Mappers() map[string][]mapper {
+	return map[string][]mapper{
+		"google_compute_firewall":               {{convert: GetComputeFirewallCaiObject}},
+		"google_compute_disk":                   {{convert: GetComputeDiskCaiObject}},
+		"google_compute_forwarding_rule":        {{convert: GetComputeForwardingRuleCaiObject}},
+		"google_compute_global_forwarding_rule": {{convert: GetComputeGlobalForwardingRuleCaiObject}},
+		"google_compute_instance":               {{convert: GetComputeInstanceCaiObject}},
+		"google_compute_network":                {{convert: GetComputeNetworkCaiObject}},
+		"google_compute_subnetwork":             {{convert: GetComputeSubnetworkCaiObject}},
+		"google_storage_bucket":                 {{convert: GetStorageBucketCaiObject}},
+		"google_sql_database_instance":          {{convert: GetSQLDatabaseInstanceCaiObject}},
+		"google_container_cluster":              {{convert: GetContainerClusterCaiObject}},
+		"google_container_node_pool":            {{convert: GetContainerNodePoolCaiObject}},
+		"google_bigquery_dataset":               {{convert: GetBigQueryDatasetCaiObject}},
+		"google_spanner_instance":               {{convert: GetSpannerInstanceCaiObject}},
+		"google_project_service":                {{convert: GetServiceUsageCaiObject}},
+		"google_pubsub_subscription":            {{convert: GetPubsubSubscriptionCaiObject}},
+		"google_pubsub_topic":                   {{convert: GetPubsubTopicCaiObject}},
+		"google_kms_crypto_key":                 {{convert: GetKMSCryptoKeyCaiObject}},
+		"google_kms_key_ring":                   {{convert: GetKMSKeyRingCaiObject}},
+		"google_filestore_instance":             {{convert: GetFilestoreInstanceCaiObject}},
+        <% products.each do |product| -%>
+            <% (product[:definitions].objects || []).each do |object| -%>
+                <% if object.min_version.name == "ga" -%>
+                    <% if !object.exclude && !object.exclude_validator -%>
+                        <% name = object.filename_override || object.name.underscore -%>
+                        <% tf_resource = object.__product.name.underscore+ "_" + name.gsub(" ", "_") -%>
```

Comment:
It looks like to be safest it might be worth matching the current logic from [`provider.go.erb`](https://github.com/GoogleCloudPlatform/magic-modules/blob/197cb1825de5facbf7b44aa58cba1288e8fa3eef/mmv1/third_party/terraform/utils/provider.go.erb#L327), assuming you have access to `config`:

```rb
tf_product = (config.legacy_name || product_definition.name).underscore
terraform_name = object.legacy_name || "google_#{tf_product}_#{object.name.underscore}"
```